### ISSUE TITLE
Why dataloader worker is 0

### ISSUE THREAD

**Issue by seanzhuh** [ultralytics/ultralytics#22611]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I'm using the following code snippet to start training of yolo-world model
```
model.train(data=data, 
            batch=256, 
            epochs=100, 
            device=[i for i in range(args.gpus)],
            amp=True,
            workers=2, # per rank
            cache=False, # "ram" takes ~0.5h for gqa caching
            compile=False,
            imgsz=640,
            split='minival',
            max_det=1000,
            save_json=True,
            project="/home/lijie/proj/xmu/sean/ultralytics/run",
            trainer=WorldTrainerFromScratch
    )
```
However I noticed that in the log, ultralytics reports that using 0 dataloader workers, what could be the problem? Running yolo check gives version 8.3.225 python-3.11.11 torch-2.9.0+cu128. Did ultralytics support H200 gpus? This problem only occurs on H200 gpu but not on GTX 4090 gpus.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @seanzhuh, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will also assist you here soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Exact code used (you’ve shared the snippet, thanks!), full logs showing “0 dataloader workers,” and any warnings/errors
-Your environment details: OS, Python, PyTorch, CUDA/cuDNN versions, GPU count, and how you’re launching training (single-GPU, multi-GPU/DDP)
-Ultralytics and model versions, and whether you’re modifying trainers (e.g., `WorldTrainerFromScratch`) or configs
-Any relevant dataset info and settings (e.g., `workers`, `batch`, `imgsz`, caching), plus a brief note on whether the same config works with a standard YOLO11 trainer
Also, if you can isolate the behavior to a short script or config using the default datasets, that would be ideal.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by seanzhuh** [comment]
It turns out that on another machine, the os.cpu_count() is 1, resulting [nw equals to 0](https://github.com/ultralytics/ultralytics/blob/4ac2d6152afbdf1d0c39649e6f6d4f7e299f1fa3/ultralytics/data/build.py#L315)

**Comment by glenn-jocher** [comment]
Good catch—this is expected from our worker heuristic in `build_dataloader`, which sets `nw = min(os.cpu_count() // max(torch.cuda.device_count(), 1), workers)`; with 1 vCPU and multiple visible GPUs the integer division yields 0, so either allocate more CPU cores, limit visible GPUs per process (DDP with one GPU per proc), or run with `workers=0` on that host; H200 is supported as long as PyTorch sees it. See the logic in the [build_dataloader reference](https://docs.ultralytics.com/reference/data/build/#build_dataloader).
